ci/version.py

```diff
@@ -0,0 +1,178 @@
+            test_version = "2.7.1"
+        set_version(current_version, test_version, dry_run=True)
+        sys.exit(0)
+
+    if len(sys.argv) == 2 and sys.argv[1] == "update":
+        print("Change all the version numbers in the repository, EXCEPT the")
+        print("current_version file in this directory, such that git diff")
+        print("will give a valid diff (make sure the repo is clean before")
+        print("doing this). The new version doesn't matter; it just serves")
+        print("as a marker and will be reverted. When done, enter the version")
+        print("you changed the current version into here.")
+        print()
+        dummy_version = input("New/dummy version: ").strip()
+        print()
+        if dummy_version == current_version:
+            print("Current and new/dummy version must differ")
+            sys.exit(1)
+
+        # Ask git for the diff, expected to correspond to a version change from
+        # current_version to dummy_version.
+        output = subprocess.run(["git", "diff"], capture_output=True)
+        if output.returncode:
+            error = output.stderr.decode("utf-8")
+            print(f"Failed to get diff:\n{error}")
+            sys.exit(1)
+        diff = output.stdout.decode("utf-8")
+
+        # Parse the diff and convert it to a format-style template.
+        lines = diff.strip().split("\n")
+        template_lines = []
+        num_changes = 0
+        skip = False
+        for index, line in enumerate(lines):
+            if skip:
+                skip = False
+                continue
+            if line.startswith("-") and not line.startswith("---"):
+                if index == len(lines) - 1:
+                    print("Error: - line at end of `git diff`")
+                    sys.exit(1)
+                frm = line[1:]
+                if not lines[index + 1].startswith("+"):
+                    print(
+                        f"Error: expecting -+ line pairs on lines {index+1} "
+                        f"and {index+2} of `git diff`"
+                    )
+                    sys.exit(1)
+                to = lines[index + 1][1:]
+                if (
+                    frm.replace(current_version, dummy_version) != to
+                    or to.replace(dummy_version, current_version) != frm
+                ):
+                    print(
+                        f"Error: diff at lines {index+1} and {index+2} of git "
+                        f"diff is not just a change from {current_version} to "
+                        f"{dummy_version}"
+                    )
+                    sys.exit(1)
+                template_lines.append(
+                    "-" + escape(frm).replace(current_version, "{frm}")
+                )
+                template_lines.append("+" + escape(to).replace(dummy_version, "{to}"))
+                skip = True
+                num_changes += 1
+            else:
+                template_lines.append(escape(line))
```

Comment:
Is all of this code doing something other than bumping the crates in the workspace?

Reply:
See https://github.com/substrait-io/substrait-validator/blob/main/RELEASE.md#incrementing-version-numbers. tl;dr, various duplicates of the version number in the various cargo.toml files, a reference in Rust's readme/docs for the customary copypasta to include a library as a dependency, and importantly pyproject.toml. It also helps update and sanity-check the patch file, which would be a massive pain otherwise.

Note that this is not exactly my idea of a clean solution, but (Python script implementation aside) I'm not sure how to do it better without making the whole thing massively more complex by special-casing everything or generating all these files.